hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpBindJoinSPARQLwithBoundJoin.java

```diff
@@ -0,0 +1,308 @@
+		}
+
+		// Generate the UNION pattern by iterating over all input solution mappings
+		// and replacing the renamed variable
+		int i = 0;
+		for ( final Binding solMap : solMaps ) {
+			// Create a new pattern in which we replace the variables
+			// with values based on the incoming solution mapping
+			SPARQLGraphPattern patternWithBindings;
+			try {
+				patternWithBindings = query.applySolMapToGraphPattern( new SolutionMappingImpl(solMap) );
+			} catch ( VariableByBlankNodeSubstitutionException e ) {
+				throw new ExecOpExecutionException(e, this);
+			}
+			
+			// Create new variable 
+			final Var v = Var.alloc( renamedVar.getVarName() + "_" + i );
+			
+			// Rename the variable in the pattern
+			final Element elt2 = renameVar(patternWithBindings, renamedVar, v);
+			union.addElement(elt2);
+			
+			// Add the solution mapping to the list of solution mappings
+			solMapsList.add(solMap);
+			// Map the renamed var to the index of the corresponding solution mapping
+			renamedVars.put(v, i);
+			i++;
+		}
+		return union;
+	}
+
+	@Override
+	protected MyIntermediateResultElementSink createMySink() {
+		if (useOuterJoinSemantics) {
+			return new MyIntermediateResultElementSinkOuterJoin2(currentBatch);
+		} else {
+			return new MyIntermediateResultElementSink2(currentBatch);
+		}
+	}
+
+	// ------- helper classes ------
+
+	protected class MyIntermediateResultElementSink2 extends MyIntermediateResultElementSink
+	{
+		protected final Set<SolutionMapping> inputSolutionMappingsWithoutJoinPartners = new HashSet<>();
+
+		public MyIntermediateResultElementSink2( final Iterable<SolutionMapping> inputSolutionMappings ) {
+			super(inputSolutionMappings);
+		}
+
+		protected void _send( final SolutionMapping smFromRequest ) {
+			// Merge smFromRequest into the input solution mappings:
+			// 1. For each renamed variable, check if the request binding contains it.
+			// 2. Rename that variable and merge with the corresponding solMapList entry.
+			// 3. For each compatible input mapping, merge and collect the results.
+			for( final Entry<Var, Integer> entry : renamedVars.entrySet() ){
+				final Var v = entry.getKey();
+				final int i = entry.getValue();
+				final Binding b = smFromRequest.asJenaBinding();
+
+				if ( ! b.contains(v) ) {
+					continue;
+				}
+
+				// Rename var and merge with input mapping
+				final Binding renamedAndMerged = BindingLib.merge( renameVar(b, v, renamedVar),
+				                                                             solMapsList.get(i) );
+				final SolutionMapping updatedRequest = new SolutionMappingImpl(renamedAndMerged);
+
+				// Merge with inputSolutionMappings
+				for ( final SolutionMapping smFromInput : inputSolutionMappings ) {
+					if ( SolutionMappingUtils.compatible(smFromInput, updatedRequest ) ) {
+						solMapsForOutput.add( SolutionMappingUtils.merge(smFromInput, updatedRequest) );
+					}
+				}
```

Comment:
I don't think that this is necessary in your case. You have done the join already. Therefore, all that remains at this point is the following.
```suggestion
				solMapsForOutput.add(updatedRequest);
```

Reply:
`BaseForExecOpBindJoinWithRequestOps#_processJoinableInput` only adds the restricted solution mappings to `currentSolMapsForRequest` (i.e., the `renamedAndMerged` only contains the join variables), which is why I had to do the join "again". However, this is probably not what we should be doing. I'll override `_processJoinableInput` to add the full input mappings instead, although this means that we no longer filer based on the restricted bindings.

An alternative design would be to restrict the solution mappings used in the request, as in the original implementation, but maintain a mapping from each request solution to the set of input solution mappings it corresponds to. This approach would introduce an additional data structure to track these associations, but would have the advantage of reducing the size of both the generated query and the response size.